web/server/go.md

```diff
@@ -0,0 +1,498 @@
+It also allows for isolating experiments and changes
+to a small portion of the codebase,
+rather than propagated them throughout all model usages.
+
+#### Where/When to Write Query Code
+
+The above examples make use of an abstracted dependency for a DB connector.
+In reality,
+any ORM or query interface is likely to be much more complex than `Fetch`/`Save`.
+As data requirements become more complex,
+it becomes much harder to maintain models as purely structural.
+It's hard to make an exact recommendation for Go code structure
+as data requirements may vary considerably.
+Backends could vary from REST APIs,
+highly normalized SQL tables,
+serialized document stores,
+sparsely column oriented data,
+and so on.
+
+Some general considerations when organizing query code are:
+
+* If query patterns vary greatly for individual models,
+  consider tying them closer to the business logic they mirror
+  (such as a service layer).
+* If individual query patterns satisfy multiple models,
+  consider a separate query layer that can abstract/share code.
+  When designing something like this,
+  be wary of over-abstracting and reinventing a query language/ORM.
+* If models are single use (ie. they are queried and fetched in one way),
+  adding query code to the model may make sense.
+  Note though, models are often shared across multiple code layers,
+  and due to struct field marshalling,
+  cannot utilize interface abstractions.
+  Keeping models as simple as possible (ie. without save/fetch methods),
+  keeps code changes from propagating across the codebase
+  and becoming unmaintainable.
+
+#### Testing Models
+
+Theoretically models should have little responsibility
+other than struct definition.
+Any exported methods should be unit tested,
+such as data validations
+or state change methods, like above.
+
+#### Model References
+
+* [SQL Standard Library](https://golang.org/pkg/database/sql/)
+  and some [helpful examples](https://golang.org/src/database/sql/example_test.go)
+* [Illustrated Guide to SQLX](https://jmoiron.github.io/sqlx/)
+
+### Services
+
+So, we have handlers,
+which delegate a request/reponse flow
+and models that describe data.
+Where does the application logic go?
+Into the service layer!
+The service layer is where work gets delegated from a handler.
+The pattern is borrowed from other languages
+[like Ruby](https://medium.com/selleo/essential-rubyonrails-patterns-part-1-service-objects-1af9f9573ca1),
+where controllers and models
+get encumbered with multiple responsibilities.
+The service layer becomes a space
+for single use, reusable objects (service objects)
+to be built out.
+This layer will often be referred to as "business logic"
+or "application logic",
+but it's more or less
+what makes your application/API useful to a client.
+It's also where the various package dependencies
+of the application
+are likely to meet,
+answering the question of
+"how do I accomplish the client need with the given dependencies?"
```

Comment:
Yes!

Also, I generally lump in the following responsibilities in the services layer: biz logic security and auth; coordination of multiple resources; etc.

Another note on security - this is the layer to be answering the question "does this principal have auth to access change this data", as opposed to doing it in the data-access layer. In my experience the data-access layer should be, and I believe by necessity, has full access to all (or at least most) of the data.

Again, I'm open to pushback on this tho!

Reply:
Also, transport-layer stuff shouldn't make it down to this layer! No `http` package things should go in/out of this layer.

Reply:
Oh, also! I would suggest all/most methods at these layers take a `context.Context` object. (Maybe this is even something that should be more generally mentioned?)

Reply:
Ahhh, maybe the `context.Context` should be in the "General Patterns" section...?

Reply:
➕ to keeping `http` packages out of here. 

Authorization always seemed to go here in my head, since it is independent of the data access method. I think the next iteration of this could talk about authorization patterns. I don't have the best pattern right now, but I know MilMove is doing some work with roles and it seems like it could relate.